tsl/src/continuous_aggs/planner.c

```diff
@@ -0,0 +1,428 @@
+
+			context->watermark_functions = lappend(context->watermark_functions, funcExpr);
+
+			/* Only on time based hypertables, we have a to_timestamp function */
+			if (context->parent_to_timestamp_func != NULL)
+			{
+				/* to_timestamp functions take only one parameter. This should be a reference to our
+				 * function */
+				Assert(linitial(context->parent_to_timestamp_func->args) == node);
+
+				context->watermark_parent_functions =
+					lappend(context->watermark_parent_functions, context->parent_to_timestamp_func);
+			}
+			else
+			{
+				/* For non int64 partitioned tables, the watermark function is wrapped into a cast
+				 * for example: COALESCE((_timescaledb_functions.cagg_watermark(11))::integer,
+				 * '-2147483648'::integer))
+				 */
+				Node *coalesce_arg = linitial(context->parent_coalesce_expr->args);
+				if (coalesce_arg != node)
+				{
+					/* Check if the watermark function is wrapped into a cast function */
+					if (!IsA(coalesce_arg, FuncExpr) || ((FuncExpr *) coalesce_arg)->args == NIL ||
+						linitial(((FuncExpr *) coalesce_arg)->args) != node)
+					{
+						context->valid_query = false;
+						return false;
+					}
+
+					context->watermark_parent_functions =
+						lappend(context->watermark_parent_functions, coalesce_arg);
+				}
+				else
+				{
+					context->watermark_parent_functions =
+						lappend(context->watermark_parent_functions, context->parent_coalesce_expr);
+				}
+			}
+		}
+
+		/* Capture the timestamp conversion function */
+		if (list_member_oid(context->to_timestamp_func_oids, funcExpr->funcid))
+		{
+			FuncExpr *old_func_expr = context->parent_to_timestamp_func;
+			context->parent_to_timestamp_func = funcExpr;
+			bool result = expression_tree_walker(node, constify_cagg_watermark_walker, context);
+			context->parent_to_timestamp_func = old_func_expr;
+
+			return result;
+		}
+	}
+	else if (IsA(node, Query))
+	{
+		/* Recurse into subselects */
+		Query *query = castNode(Query, node);
+		return query_tree_walker(query,
+								 constify_cagg_watermark_walker,
+								 context,
+								 QTW_EXAMINE_RTES_BEFORE);
+	}
+	else if (IsA(node, CoalesceExpr))
+	{
+		/* Capture the CoalesceExpr */
+		CoalesceExpr *parent_coalesce_expr = context->parent_coalesce_expr;
+		context->parent_coalesce_expr = castNode(CoalesceExpr, node);
+		bool result = expression_tree_walker(node, constify_cagg_watermark_walker, context);
+		context->parent_coalesce_expr = parent_coalesce_expr;
+
+		return result;
+	}
+	else if (IsA(node, RangeTblEntry))
+	{
+		/* Collect the Oid of the used range tables */
+		RangeTblEntry *rte = (RangeTblEntry *) node;
+
+		if (rte->rtekind == RTE_RELATION)
+		{
+			context->relids = list_append_unique_oid(context->relids, rte->relid);
+		}
+
+		/* allow range_table_walker to continue */
+		return false;
+	}
+
+	return expression_tree_walker(node, constify_cagg_watermark_walker, context);
+}
+
+/*
+ * Check if the given query is a union query
+ */
+static bool
```

Comment:
```suggestion
static inline bool
```
What about in-lining it?

Reply:
Good point, changed